### PR TITLE
Add support LogpullReceived and LogpullFields.

### PR DESCRIPTION
**Pull request by cyriltovena**
## Description

Add support `LogpullReceived` and `LogpullFields` as described by https://developers.cloudflare.com/logs/logpull/requesting-logs

This PR allows to pull logs and logs field from the log pull API. The `LogpullReceived` returns a streaming iterator that can iterates over line for fields value depending on your need.

I choose to always use gzip for `LogpullReceived` as logs are highly compressible and will help in reducing network load.

I had to also add support for:
- Gzip response content encoding
- Support for getting reader from body to avoid reading everything into a byte slice and gzip decode along the way.
- Non-json response errors, required for `LogpullReceived`.

## Has your change been tested?

Tested via unit testing and using cloudflare API. 
I found out that some response are not in JSON and update the code to work accordingly.

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented (api.cloudflare.com or developers.cloudflare.com) and stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/

### CODE REVIEW

File: logpull_test.go [hunk 1]
```diff
@@ -60,3 +63,85 @@ func TestSetLogpullRetentionFlag(t *testing.T) {
...
+		assert.Equal(t, http.MethodGet, r.Method, "Expected method 'GET', got %s", r.Method)
+		w.Header().Set("content-type", "application/json")
+		fmt.Fprintf(w, `{
+		"foo": "bar",
+		"field_1": "value_1",
+		"field_2": "value_2"
+	}`)
+	}
+
+	mux.HandleFunc("/zones/e49c6317afd4ceabe24c6e11d2dad231/logs/received/fields", handler)
```
Comment (buro9) [inline logpull_test.go:81]: Remove hard-coded zoneID

File: logpull_test.go [hunk 2]
```diff
@@ -60,3 +63,85 @@ func TestSetLogpullRetentionFlag(t *testing.T) {
...
+
+		w.Header().Set("Content-Encoding", "gzip")
+		wg := gzip.NewWriter(w)
+		for _, line := range want {
+			fmt.Fprintln(wg, line)
+		}
+		assert.NoError(t, wg.Close())
+	}
+
+	mux.HandleFunc("/zones/e49c6317afd4ceabe24c6e11d2dad231/logs/received", handler)
```
Comment (buro9) [inline logpull_test.go:125]: Remove hard-coded zoneID

File: logpull_test.go [hunk 3]
```diff
@@ -60,3 +63,85 @@ func TestSetLogpullRetentionFlag(t *testing.T) {
...
+
+	l, err := client.LogpullReceived(context.Background(), "e49c6317afd4ceabe24c6e11d2dad231", start, end, LogpullReceivedOption{
```
Comment (buro9) [inline logpull_test.go:127]: Remove hard-coded zoneID

**Comment by buro9** [comment]
@twalwyn and @mkocikowski , Grafana is adding support to the Cloudflare SDK (this repo) to support reading logs as the basis for then pushing those logs into [Loki](https://github.com/grafana/loki/) . I notice that Jacob (usual maintainer of this repo) is on Paternity Leave (according to Github status). As you two are best placed to comment or point a Cloudflare engineer to comment, could we get a steer on whether you are happy with the approach we are taking here and the implementation. This is likely the first of several PRs which will fully implement reading logs via the Cloudflare SDK.

**Comment by jacobbednarz** [comment]
thanks for being so considerate @buro9 🤗 I really appreciate it. 

@patryk will be looking after majority of my duties while I'm out however it is in a very limited adhoc capacity for critical issues. this does however sound like a great partnership so it might be worth chatting with @garrettgalow to get someone allocated to review and assist this through the pipeline. I'm back full time in January but will be spotty over Christmas if this still needs some attention.

File: cloudflare.go [hunk 4]
```diff
@@ -270,19 +278,102 @@ func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, u
...
+type closeDiscardBody struct {
+	body io.ReadCloser
+}
+
+func (b *closeDiscardBody) Read(p []byte) (n int, err error) {
+	return b.body.Read(p)
+}
+
+func (b *closeDiscardBody) Close() error {
+	_, _ = io.Copy(ioutil.Discard, b.body)
```
Comment (victor-perov) [inline cloudflare.go:319]: Despite that here we use `ioutil.Discard` (that will succeed on any write), If `b.body.Read()` fails, the error will be returned. Should we care about this scenario?
```suggestion
	_, err := io.Copy(ioutil.Discard, b.body)
	if err != nil {
		return errors.Wrap(err, "failed to discard response body")
	}
```

File: logpull.go [hunk 5]
```diff
@@ -57,3 +63,113 @@ func (api *API) SetLogpullRetentionFlag(ctx context.Context, zoneID string, enab
...
+	if err != nil {
+		return nil, err
+	}
+	return &logpullReceivedResponse{
+		scanner: bufio.NewScanner(reader),
+		reader:  reader,
+	}, nil
+}
+
+func (r *logpullReceivedResponse) Next() bool {
```
Comment (victor-perov) [inline logpull.go:137]: The exported function(s) should have a comment describing its purpose

**Review by victor-perov — COMMENTED** [review]
Thanks for your work! Just some comments from me about error handling.

**Comment by tanushreesharma-cf** [comment]
Hi @buro9, 

I’m the Product Manager for Logs at Cloudflare. Would love to work with you to get an integration up. It’s something our customers have requested in the past and will be excited about! 

We should take a step back here though, I don’t think that integrating with Logpull is the best approach. Let's up some time to chat through it and discuss options. 

Shoot me an email (tanushree@cloudflare.com) and we can set up a meeting. Thanks!

**Comment by cyriltovena** [comment]
@victor-perov I've took care of your feedback, let me know if there is anything else.

Cheers.

**Comment by tanushreesharma-cf** [comment]
Hey there @cyriltovena! See my comment above.

I can see that you've been added to the email thread I have with the Grafana team. Let's put a hold on this until we get a chance to speak.

**Comment by jacobbednarz** [comment]
@tanushreesharma-cf @cyriltovena do we still require this PR? if so, what do we need to do to get this over the line?

**Comment by tanushreesharma-cf** [comment]
We've spoken to the Grafana Loki team and they'll be building out an integration with Logpush so I don't think this is needed from their side anymore.

**Comment by jacobbednarz** [comment]
thanks! I'll close this out but we can always revive it if we need it.

**Pull request closed without merging**